materialize-bigquery/sqlgen.go

```diff
@@ -0,0 +1,233 @@
+	typeMapper := sql.NullableMapper{
+		NotNullText: "NOT NULL",
+		Delegate:    bqTypeMapper,
+	}
+
+	return sql.Dialect{
+		Identifierer: sql.IdentifierFn(sql.JoinTransform(".",
+			sql.PassThroughTransform(
+				func(s string) bool {
+					return sql.IsSimpleIdentifier(s)
+				},
+				identifierSanitizer,
+			))),
+		Literaler: sql.LiteralFn(sql.QuoteTransform("'", "''")),
+		Placeholderer: sql.PlaceholderFn(func(_ int) string {
+			return "?"
+		}),
+		TypeMapper: typeMapper,
+	}
+}()
+
+var (
+	tplAll = sql.MustParseTemplate(bqDialect, "root", `
+{{ define "tempTableName" -}}
+flow_temp_table_{{ $.Binding }}
+{{- end }}
+
+-- Templated creation of a materialized table definition and comments:
+
+{{ define "createTargetTable" -}}
+CREATE TABLE IF NOT EXISTS {{$.Identifier}} (
+	{{- range $ind, $col := $.Columns }}
+		{{- if $ind }},{{ end }}
+		{{$col.Identifier}} {{$col.DDL}}
+	{{- end }}
+)
+CLUSTER BY {{ range $ind, $key := $.Keys }}
+	{{- if $ind }}, {{end -}}
+		{{$key.Identifier}}
+	{{- end}};
+{{ end }}
+
+-- Templated query which joins keys from the load table with the target table,
+-- and returns values. It deliberately skips the trailing semi-colon
+-- as these queries are composed with a UNION ALL.
+
+{{ define "loadQuery" -}}
+{{ if $.Document -}}
+SELECT {{ $.Binding }}, l.{{$.Document.Identifier}}
+	FROM {{ $.Identifier }} AS l
+	JOIN {{ template "tempTableName" . }} AS r
+	{{- range $ind, $key := $.Keys }}
+		{{ if $ind }} AND {{ else }} ON {{ end -}}
+		l.{{ $key.Identifier }} is not distinct from r.{{ $key.Identifier }}
+	{{- end }}
+{{ else }}
+SELECT -1, NULL LIMIT 0
+{{ end }}
+{{ end }}
+
+-- Templated query which bulk inserts rows from a source bucket into a target table.
+
+{{ define "storeInsert" -}}
+INSERT INTO {{ $.Identifier }} (
+	{{- range $ind, $col := $.Columns }}
+		{{- if $ind }}, {{ end -}}
+		{{$col.Identifier}}
+	{{- end -}}
+)
+SELECT {{ range $ind, $col := $.Columns }}
+		{{- if $ind }}, {{ end -}}
+		{{$col.Identifier}}
+	{{- end }} FROM {{ template "tempTableName" . }};
+{{ end }}
+
+-- Templated query which updates an existing row in the target table:
+
+{{ define "storeUpdate" -}}
+MERGE INTO {{ $.Identifier }} AS l
+USING {{ template "tempTableName" . }} AS r
+ON {{ range $ind, $key := $.Keys }}
+{{- if $ind }} AND {{end -}}
+	l.{{$key.Identifier}} = r.{{$key.Identifier}}
+{{- end}}
+WHEN MATCHED AND r.{{$.Document.Identifier}} IS NULL THEN
+	DELETE
+WHEN MATCHED THEN
+	UPDATE SET {{ range $ind, $key := $.Values }}
+	{{- if $ind }}, {{end -}}
+		l.{{$key.Identifier}} = r.{{$key.Identifier}}
+	{{- end}} 
+	{{- if $.Document }}, l.{{$.Document.Identifier}} = r.{{$.Document.Identifier}} {{- end }}
+WHEN NOT MATCHED THEN
+	INSERT (
+	{{- range $ind, $col := $.Columns }}
+		{{- if $ind }}, {{ end -}}
+		{{$col.Identifier}}
+	{{- end -}}
+	)
+	VALUES (
+	{{- range $ind, $col := $.Columns }}
+		{{- if $ind }}, {{ end -}}
+		r.{{$col.Identifier}}
+	{{- end -}}
+	);
+{{ end }}
+
+{{ define "installFence" }}
```

Comment:
This fence installation query was copied as-is from the existing code. I think there may be opportunities to simplify the query, but wasn't able to readily identify anything significant. It may be an are for future improvement. The current query does work correctly.